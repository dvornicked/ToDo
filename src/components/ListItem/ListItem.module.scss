@use "sass:color";

.list-item {
  margin: 10px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.text {
  font-size: 1.5rem;
  border-bottom: 1px steelblue dashed;
  flex-grow: 1;
  cursor: pointer;
}

.important {
  font-weight: bold;
}

.done {
  text-decoration: line-through;
}

.button {
  background: none;
  border: 1px gray solid;
  border-radius: 5px;
  margin-left: 10px;
  cursor: pointer;

  svg {
    margin-top: 2px;
  }

  &.toggled, &.exclamation:hover {
    background: color.adjust(steelblue, $lightness: 10%);

    svg {
      fill: white;
    }
  }

  &.trash:hover {
    background: color.adjust(red, $lightness: 20%);

    svg {
      fill: white;
    }
  }
}
