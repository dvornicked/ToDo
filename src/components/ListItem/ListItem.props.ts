import { DetailedHTMLProps, HTMLAttributes } from 'react'

export interface IFuncTask {
  deleteTask(idx: number): void
  setImportant(idx: number): void
  setDone(idx: number): void
}

export interface ITask {
  text: string
  important?: boolean
  done?: boolean
}

export interface ListItemProps extends Omit<
DetailedHTMLProps<HTMLAttributes<HTMLLIElement>, HTMLLIElement>,
'onAnimationStart' | 'onDragStart' | 'onDragEnd' | 'onDrag' | 'ref'
> {
  task: ITask
  funcTask: IFuncTask
  idx: number
}
