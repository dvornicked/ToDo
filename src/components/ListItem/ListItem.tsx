/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import cn from 'classnames'
import { motion } from 'framer-motion'
import { ListItemProps } from './ListItem.props'
import styles from './ListItem.module.scss'
import { ReactComponent as Trash } from './trash.svg'
import { ReactComponent as Exclamation } from './exclamation.svg'

function ListItem({
  task: { text, important = false, done = false },
  funcTask: { deleteTask, setDone, setImportant },
  idx,
  className,
  ...props
}: ListItemProps) {
  const handleKeyDown = () => {}
  return (
    <motion.li
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={{ opacity: 0 }}
      className={cn(styles['list-item'], className)}
      {...props}
    >
      <span
        className={cn(styles.text, {
          [styles.important]: important,
          [styles.done]: done,
        })}
        onClick={() => setDone(idx)}
        onKeyDown={handleKeyDown}
        role="term"
      >
        {text}
      </span>
      <button
        type="button"
        className={cn(styles.button, styles.exclamation, {
          [styles.toggled]: important,
        })}
        title="Exclamation"
        onClick={() => {
          setImportant(idx)
        }}
      >
        <Exclamation />
      </button>
      <button
        type="button"
        className={cn(styles.button, styles.trash)}
        title="Trash"
        onClick={() => deleteTask(idx)}
      >
        <Trash />
      </button>
    </motion.li>
  )
}

export default ListItem
