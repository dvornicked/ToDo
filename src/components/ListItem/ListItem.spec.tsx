import { render, screen } from '@testing-library/react'

import ListItem from './ListItem'
import { ITask } from './ListItem.props'

const task: ITask = { text: 'Task 1' }
const funcTask = {
  setImportant: jest.fn(),
  setDone: jest.fn(),
  deleteTask: jest.fn(),
}

describe('ListItem component', () => {
  it('ListItem renders', () => {
    render(<ListItem funcTask={funcTask} task={task} idx={0} />)

    expect(screen.getByText('Task 1')).toBeInTheDocument()
  })

  it('ListItem snapshot', () => {
    const view = render(<ListItem funcTask={funcTask} task={task} idx={0} />)
    expect(view).toMatchSnapshot()
  })
})
