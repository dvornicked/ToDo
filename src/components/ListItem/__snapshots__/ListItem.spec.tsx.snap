// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component ListItem snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="list-item"
        style="opacity: 0;"
      >
        <span
          class="text"
          role="term"
        >
          Task 1
        </span>
        <button
          class="button exclamation"
          title="Exclamation"
          type="button"
        >
          <svg>
            exclamation.svg
          </svg>
        </button>
        <button
          class="button trash"
          title="Trash"
          type="button"
        >
          <svg>
            trash.svg
          </svg>
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="list-item"
      style="opacity: 0;"
    >
      <span
        class="text"
        role="term"
      >
        Task 1
      </span>
      <button
        class="button exclamation"
        title="Exclamation"
        type="button"
      >
        <svg>
          exclamation.svg
        </svg>
      </button>
      <button
        class="button trash"
        title="Trash"
        type="button"
      >
        <svg>
          trash.svg
        </svg>
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
